### ISSUE TITLE
Export YOLOv8 model to "engine" yields a very small ONNX file & fails

### ISSUE THREAD

**Issue by MrMaynard** [ultralytics/ultralytics#1833]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have built a model and exported it:
```
best.pt
```

I then try and load this model and write it to ONNX:
```
model = YOLO("best.pt")
model.export(format="onnx", device=0, imgsz=400)
```

This works fine and results in an ONNX model about 50MB, the same size as my .pt model.

However, I want to export to TensorRT. I have installed CUDA 11.7 and TensorRT 8.5.3.1 on Windows 10. I run the following:
```
model = YOLO("best.pt")
model.export(format="engine", device=0, imgsz=400)
```

This results in a very small ONNX file being written, about 1 MB. I see output like this in the console:
```
PyTorch: starting from runs\detect\train\weights\best.pt with input shape (1, 3, 416, 416) BCHW and output shape(s) (1, 6, 3549) (49.6 MB)
ONNX: starting export with onnx 1.13.1...
ONNX: simplifying with onnxsim 0.4.19...
ONNX: export success  2.9s, saved as runs\detect\train\weights\best.onnx (0.4 MB)
TensorRT: starting export with TensorRT 8.5.3.1...
```

Later, the export fails on this ONNX file:
```
[04/05/2023-01:44:40] [TRT] [E] ModelImporter.cpp:732: ERROR: builtin_op_importers.cpp:697 In function importConv:
[8] Assertion failed: (nbSpatialDims == kernelWeights.shape.nbDims - 2) && "The number of spatial dimensions and the kernel shape doesn't match up for the Conv operator."
TensorRT: export failure  5.6s: failed to load ONNX file: runs\detect\train3\weights\best.onnx
```

What could be causing the ONNX export to fail, and later the export to TensorRT to fail? Thanks.

**Comment by glenn-jocher** [comment]
Hello @0xdeadface, it seems like the ONNX export is failing to simplify using `onnxsim`, which could be the result of a couple of things. 

One thing you could try is to reducing the `imgsz` parameter to 320 or 288, as sometimes a larger size can cause complexity issues with ONNX that can propagate through to TensorRT.

You may also want to try using an earlier version of TensorRT, such as 7.2, as TensorRT 8.5.3 can sometimes have issues with certain models. Additionally, it may be worth checking that your GPU architecture is supported by your version of TensorRT.

Hope this helps!

**Comment by MrMaynard** [comment]
Thanks @glenn-jocher for the help. 

Unfortunately adjusting the `imgsz` did not help. I tried lowering my TensorRT to 7.2.3, but I got this when trying to export:
```
TensorRT: export failure  10.1s: Could not find module 'C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.7\bin\nvinfer.dll' (or one of its dependencies). Try using the full path with constructor syntax.
```
I suppose this might be because CUDA 11.7 isn't supported by TensorRT 7.2?

**Comment by MrMaynard** [comment]
Small update...

I tried lowering my CUDA to 11.2, my PyTorch to 1.8.0+cu111 , my TensorRT to 7.2.3, and my cuDNN to 8.1.1, but I must have an incompatibility somewhere as the export fails:

```
TensorRT: export failure  0.6s: DLL load failed while importing tensorrt: The specified procedure could not be found.
```

**Comment by MrMaynard** [comment]
Right, but I'm only seeing that error messages now that you've explicitly suggested I use TensorRT 7.2. If I use the latest YOLOv8 and the latest CUDA and the latest TensorRT there is no version incompatibility. YOLOv8's export nonetheless fails as initially reported. You suggested I try rolling back some dependency versions as a workaround, but is this really not a YOLOv8 issue?

**Comment by MrMaynard** [comment]
If YOLOv8 isn’t able to properly simplify the ONNX, is it possible that we can directly provide a simplified model to the engine export? I tried that, but it complained that it expected a .pt

**Comment by glenn-jocher** [comment]
@MrMaynard yes, it's possible to supply a simplified model directly to the engine export, but unfortunately it must still be in ONNX format. ONNX is an intermediate representation designed to allow models to be easily transferred between different frameworks, and it must be used by TensorRT for model inference. Therefore, even if you are able to simplify the model, it still must be exported in ONNX format. Simply renaming the file extension will not work either, as the content and structure of an ONNX model is different from that of a PyTorch model.

**Comment by evenjo** [comment]
I have the exact same issue, have looked everywhere without finding a solution.
@glenn-jocher The issue here isn't the process of converting to ONNX format, but the process after, which is simplifying the ONNX model with ONNXSIM. I have converted it to a 'normal' ONNX format, and that gives me a ~12MB file, but when trying to simplify, it ends up at 0.1MB.

**Comment by MrMaynard** [comment]
@glenn-jocher 
> @MrMaynard yes, it's possible to supply a simplified model directly to the engine export, but unfortunately it must still be in ONNX format.

As I mentioned I tried that, but YOLOv8's export fails:

```
model = YOLO("runs/detect/train/weights/simplified.onnx")
model.export(format="engine", imgsz=400, half=True, device=0)
```

>TypeError: model='runs/detect/train3/weights/simplified.onnx' must be a *.pt PyTorch model, but is a different type. PyTorch models can be used to train, val, predict and export, i.e. 'yolo export model=yolov8n.pt', but exported formats like ONNX, TensorRT etc. only support 'predict' and 'val' modes, i.e. 'yolo predict model=yolov8n.onnx'.

**Comment by MrMaynard** [comment]
I should note that I tried generating my own simplified ONNX:
```
import onnx
from onnxsim import simplify
    
model = onnx.load("bin/best.onnx")
model_simp, _= simplify(model)
onnx.save(model_simp, "bin/simplified.onnx")
```

And then converting it with `trtexec`:

```
trtexec --onnx=bin/simplified.onnx --saveEngine=bin/simplified.engine
```

I see the same error:
> [6] Assertion failed: (splitList.empty() || (splitList.size() == numOutputs)) && "The shape of the split attribute misaligns with the number of outputs."

So indeed YOLOv8 isn't doing anything special here, it's just calling into onnxsim and then trying to convert to TensorRT. I think the open question is why YOLOv8 trained a model that can't be converted and whether it's possible to convert it on the supported CUDA 11.8 and TensorRT 8.5.

**Comment by glenn-jocher** [comment]
@MrMaynard it appears that the issue is related to the simplification process with onnxsim not producing the expected output, resulting in the ONNX model not being compatible with TensorRT during the export. YOLOv8 follows standard procedures for exporting models to TensorRT and relies on external libraries like onnxsim to simplify the model. The open question is why the simplified ONNX model doesn't work, and whether there is another solution that would allow the model to be converted on the supported CUDA and TensorRT versions.

**Comment by MrMaynard** [comment]
Got it, I understand that under the covers this may be an onnxsim issue. However, I'm just using the version of onnxsim that shipped with YOLOv8. From the user perspective onnxsim is just a dependency and I think it's reasonable to assume that it will work. If YOLOv8 is able to read a .pt model and it supports writing an .engine model, then it should probably ship with a dependency that's able to do so.

**Comment by glenn-jocher** [comment]
@MrMaynard i understand your frustration with the onnxsim dependency not working as expected, despite it being listed as a dependency. However, it's important to note that the YOLOv8 repository relies on several open-source libraries, and compatibility issues can arise due to changes in versions or configurations, which can be difficult to anticipate or control. In such cases, the best approach would be to seek assistance from the community or the developers to help resolve the issue or to identify possible workarounds. In the meantime, it may be possible to export the model in a different format, like the PyTorch format or a different ONNX format, until the compatibility issue is resolved.

**Comment by MrMaynard** [comment]
Yes, note that there is an onnxsim issue [here](https://github.com/daquexian/onnx-simplifier/issues/260) which matches my description above -- I have commented there.

To the extend that YOLOv8 is a product which documents the ability to export a trained model to `engine`, whether or not some portion of that functionality is provided through a dependency is immaterial. It is YOLOv8 which documents this ability (without qualification) and so it's YOLOv8 that I would expect to provide this capability.

You've mentioned compatibility issues a number of times -- [onnxsim is pulled in as a YOLOv8 dependency](https://github.com/ultralytics/ultralytics/blob/main/requirements.txt#L28), so if there is such an issue, I would expect it to be resolved in YOLOv8. Is there a version of onnxsim where this the behavior YOLOv8 documents is known to work? I'd be happy to experiment and try another version, even if it's not the one that YOLOv8 includes as a dependency.

**Comment by glenn-jocher** [comment]
@MrMaynard I understand your concern about relying on dependencies that may have compatibility issues, and I can appreciate the expectation that YOLOv8 should provide the capability it documents. While YOLOv8's dependency on onnxsim can contribute to compatibility issues, there may be other factors at play that are outside of YOLOv8 and onnxsim's control. That being said, it may be helpful to experiment with different versions of onnxsim to see if any of them address the specific issue you're encountering.

**Comment by thusinh1969** [comment]
I would stick to Yolov5 for now, works great, and SparseML just made YoloV5 and Openvino greater performance.

YoloV8 would probably need more time. I love Yolov8 the way it simplified things, but compatibility on thing like this will take time.

**Comment by glenn-jocher** [comment]
@thusinh1969 I understand your sentiment regarding the current state of compatibility with YOLOv8, and the appeal of using YOLOv5 with SparseML for your needs. While YOLOv8 offers simplified workflows, it's important to note that compatibility and performance issues can arise with any software product. The development team is continually working to improve YOLOv8's compatibility with different platforms and configurations, and it may be worth checking back periodically for updates. In the meantime, it's great to hear that YOLOv5 and SparseML are meeting your needs.

**Comment by MrMaynard** [comment]
Yeah, I tried training a whole new YOLOv8 model from a different pretrained model and hit the same issue. From the thread here it seems like YOLOv8 is not fully-baked or at least all the documented features are not supported.

**Comment by glenn-jocher** [comment]
@MrMaynard YOLOv8 TensorRT exports works correctly and without issue. We run CI every 24 hours and on every commit, and any anomaly would be instantly caught as our CI would show red instead of green everywhere like the main README.

<img width="812" alt="Screenshot 2023-04-09 at 22 16 09" src="https://user-images.githubusercontent.com/26833433/230794576-0dace7b8-3a04-469f-a416-eefa4e98e4a2.png">

We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub with `git pull` or `pip install -U ultralytics` to ensure your problem has not already been solved in the latest code version.
- ✅ **Unmodified** – Your problem must be reproducible using official Ultrlaytics code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/ultralytics/issues/new/choose) with a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by glenn-jocher** [comment]
@MrMaynard @thusinh1969 simplifying with onnxsim also runs correctly and with no issues in my tests in Colab:

<img width="677" alt="Screenshot 2023-04-09 at 22 19 16" src="https://user-images.githubusercontent.com/26833433/230794679-668ea1ce-3284-47b2-8873-2e0fb754c880.png">

**Comment by MrMaynard** [comment]
That's a different model, though. A model that reproduces the issue is attached here.

I was able to simplify my model with YOLOv5, thanks for the suggestion @thusinh1969! @glenn-jocher the API may work for _some_ inputs but at least for my input, there's clearly a regression here.

[minified.zip](https://github.com/ultralytics/ultralytics/files/11186376/minified.zip)

**Comment by glenn-jocher** [comment]
@MrMaynard the zip you supplied contains as ONNX, but to reproduce any issues would require a YOLOv8 PyTorch model, or better yet a full training workflow with the latest version of ultralytics in a common environment like Colab. See steps for providing a minimum reproducible example at https://stackoverflow.com/help/minimal-reproducible-example

**Comment by MrMaynard** [comment]
My issue is resolved by just using YOLOv5 to avoid this regression. The ONNX file I included can be used with onnxsim like the (minimal) code snippet [I shared above illustrates](https://github.com/ultralytics/ultralytics/issues/1833#issuecomment-1498049964). If you'd prefer an .pt file, there's one attached here.

[minified.zip](https://github.com/ultralytics/ultralytics/files/11186707/minified.zip)

**Comment by glenn-jocher** [comment]
@MrMaynard I see, thank you for clarifying. I'm glad to hear that you were able to find a workaround by using YOLOv5. It sounds like the ONNX file you provided works with onnxsim, which is good to know.  We'll keep an eye out for any other reports of regression with YOLOv8 and ONNX export, and continue to work on improving compatibility with different platforms and configurations.

**Comment by thusinh1969** [comment]
> That's a different model, though. A model that reproduces the issue is attached here.
> 
> I was able to simplify my model with YOLOv5, thanks for the suggestion @thusinh1969! @glenn-jocher the API may work for _some_ inputs but at least for my input, there's clearly a regression here.
> 
> [minified.zip](https://github.com/ultralytics/ultralytics/files/11186376/minified.zip)

If you tend to run on CPU, a cheap Intel NUC or something on cloud (Intel i5 4-8 cores for example, cheaper then Jetson all the same kind of prices), export Yolov5 to Openvino, it is even faster (at twice sometimes) then sparseml re-training for both object detection and segmentation! FYI.

**Comment by glenn-jocher** [comment]
@thusinh1969 Thank you for sharing your experience and suggestion regarding using YOLOv5 and OpenVINO for improved performance in certain environments. It's important for users to consider their specific use case and hardware constraints when selecting a product, as different models and frameworks may perform differently in different environments. Additionally, evaluating the impact on accuracy is a critical consideration when comparing different models and frameworks. The Ultralytics team will continue to explore ways to improve performance and compatibility across different environments with YOLOv8.

**Comment by Fischer-pixel** [comment]
Maybe you can roll back "onnxsim" to version 0.4.0. I had the same problem with version 0.4.20, but the rollback solved it. Hope it helped.

**Comment by glenn-jocher** [comment]
@Fischer-pixel thank you for sharing your experience regarding the issue with newer versions of "onnxsim" and suggesting a workaround by rolling back to version 0.4.0. We will consider this suggestion in our ongoing work to improve compatibility and performance with YOLOv8 and related frameworks. It's valuable feedback from the community that helps us make informed decisions and prioritize our development efforts.

**Comment by MrMaynard** [comment]
@glenn-jocher per my [issue on onnxsim](https://github.com/daquexian/onnx-simplifier/issues/271) this bug is not present in onnsxim 0.4.17, for the time being i would suggest you pin the onnxsim dependency accordingly

**Comment by glenn-jocher** [comment]
@MrMaynard Decimus Meridius, thank you for sharing your experience with using onnxsim 0.4.17 and reporting the bug in onnxsim 0.4.20. We appreciate the suggestion to pin the onnxsim dependency accordingly, and we will take it into consideration in our ongoing efforts to improve compatibility and performance with YOLOv8 and related frameworks. As always, feedback from the community is valuable to inform our development efforts and ensure that we are meeting the needs of our users.

**Comment by daquexian** [comment]
Hi, I'm the author of onnxsim :) Thanks for using onnxsim and posting the bug. I have fixed it in the latest 0.4.21 version. Please have a try.

**Comment by glenn-jocher** [comment]
@daquexian Thank you for your contribution to onnxsim and for fixing the bug. It's great to see community members actively working together to improve compatibility and performance with YOLOv8 and related frameworks. We appreciate your continuous efforts to enhance the functionality and user experience of onnxsim.
